astropy/units/quantity_helper/function_helpers.py

```diff
@@ -1277,7 +1277,15 @@ def array2string(a, *args, **kwargs):
     # also work around this by passing on a formatter (as is done in Angle).
     # So, we do nothing if the formatter argument is present and has the
     # relevant formatter for our dtype.
-    formatter = args[6] if len(args) >= 7 else kwargs.get("formatter")
+    import inspect
+
+    sig = inspect.signature(np.array2string)
+    formatter_arg_pos = list(sig.parameters).index("formatter") - 1
```

Comment:
note to reviewers (which I assume would be @mhvk): I don't understand *why* this `-1` is needed, but empirically, that's what works.